src/time_series_metadata_store.jl

```diff
@@ -0,0 +1,1004 @@
+    store = TimeSeriesMetadataStore(db)
+    @debug "Loaded time series metadata from file" _group = LOG_GROUP_TIME_SERIES filename
+    return store
+end
+
+"""
+Load a TimeSeriesMetadataStore from an HDF5 file into an in-memory database.
+"""
+function from_h5_file(::Type{TimeSeriesMetadataStore}, src::AbstractString, directory)
+    data = HDF5.h5open(src, "r") do file
+        file[HDF5_TS_METADATA_ROOT_PATH][:]
+    end
+
+    filename, io = mktemp(isnothing(directory) ? tempdir() : directory)
+    write(io, data)
+    close(io)
+    return TimeSeriesMetadataStore(filename)
+end
+
+function _create_metadata_table!(store::TimeSeriesMetadataStore)
+    # TODO: SQLite createtable!() doesn't provide a way to create a primary key.
+    # https://github.com/JuliaDatabases/SQLite.jl/issues/286
+    # We can use that function if they ever add the feature.
+    schema = [
+        "id INTEGER PRIMARY KEY",
+        "time_series_uuid TEXT NOT NULL",
+        "time_series_type TEXT NOT NULL",
+        "time_series_category TEXT NOT NULL",
+        "initial_timestamp TEXT NOT NULL",
+        "resolution_ms INTEGER NOT NULL",
+        "horizon INTEGER",
+        "horizon_time_ms INTEGER",
+        "interval_ms INTEGER",
+        "window_count INTEGER",
+        "length INTEGER",
+        "name TEXT NOT NULL",
+        "owner_uuid TEXT NOT NULL",
+        "owner_type TEXT NOT NULL",
+        "owner_category TEXT NOT NULL",
+        "features TEXT NOT NULL",
+        # The metadata is included as a convenience for serialization/de-serialization,
+        # specifically for types: time_series_type and scaling_factor_multplier.
+        # There is a lot duplication of data.
+        "metadata JSON NOT NULL",
+    ]
+    schema_text = join(schema, ",")
+    _execute(store, "CREATE TABLE $(METADATA_TABLE_NAME)($(schema_text))")
+    @debug "Created time series metadata table" schema _group = LOG_GROUP_TIME_SERIES
+    return
+end
+
+function _create_indexes!(store::TimeSeriesMetadataStore)
+    # Index strategy:
+    # 1. Optimize for these user queries with indexes:
+    #    1a. all time series attached to one component/attribute
+    #    1b. time series for one component/attribute + name + type
+    #    1c. time series for one component/attribute with all features
+    # 2. Optimize for checks at system.add_time_series. Use all fields and features.
+    # 3. Optimize for returning all metadata for a time series UUID.
+    SQLite.createindex!(store.db, METADATA_TABLE_NAME, "by_id", "id"; unique = true)
+    SQLite.createindex!(store.db, METADATA_TABLE_NAME, "by_c", "owner_uuid"; unique = false)
+    SQLite.createindex!(
+        store.db,
+        METADATA_TABLE_NAME,
+        "by_c_n_tst",
+        ["owner_uuid", "name", "time_series_type"];
+        unique = false,
+    )
+    SQLite.createindex!(
+        store.db,
+        METADATA_TABLE_NAME,
+        "by_c_n_tst_features",
+        ["owner_uuid", "name", "time_series_type", "features"];
```

Comment:
Cursory Googling suggests that if you create a composite index with a bunch of columns, then put just the first few of those columns in a WHERE clause, it can still use that index — e.g., it can use `by_c_n_tst_features` if you only specify `owner_uuid`. If that's true, can we eliminate `by_c` and `by_c_n_tst`?

Reply:
Yes, you are correct. Thanks. Here is the official documentation:
```
It is not necessary for every column of an index to appear in a WHERE clause term in order for that index to be used. However, there cannot be gaps in the columns of the index that are used.
```
Under section 2.
https://www.sqlite.org/optoverview.html